File: apple/VniDrop/Core/CoreModels.swift
import Foundation

/// App-facing domain models, ported from `core/CoreModels.kt`. The repository maps
/// the generated UniFFI records/enums into these so the UI never depends on the
/// binding surface directly.

struct CoreStatus: Equatable, Sendable {
	let endpointId: String
	let activeTransfers: UInt64
	let activeShares: UInt64
}

struct CoreEventModel: Equatable, Identifiable, Sendable {
	let id: String
	let timestamp: Int64
	let scope: String
	let transferId: UInt64?
	let direction: String?
	let phase: String
	let kind: String
	let dataJson: String
}

enum ShareAccessPolicy: Equatable, Sendable {
	case requireApproval
	case anyoneWithTransfer
}

enum TransferDirection: Equatable, Sendable {
	case send
	case receive
}

enum TransferStatus: Equatable, Sendable {
	case importing
	case sharing
	case receiving
	case done
	case failed
	case cancelled
	case stopped
}

struct Transfer: Equatable, Identifiable, Sendable {
	let localId: String
	let transferId: UInt64
	let direction: TransferDirection
	let status: TransferStatus
	let peerId: String?
	let transferName: String?
	let contentHash: String?
	let fileCount: UInt64
	let totalSize: UInt64
	let ticket: String?
	let accessPolicy: ShareAccessPolicy
	let createdAt: Int64
	let updatedAt: Int64

	var id: String { localId }
}

struct Share: Equatable, Sendable {
	let transferId: UInt64
	let ticket: String
	let transferName: String
	let contentHash: String
	let fileCount: UInt64
	let totalSize: UInt64
}

struct TransferMetadataModel: Equatable, Sendable {
	let transferId: UInt64
	let transferName: String
	let senderName: String?
	let contentHash: String
	let fileCount: UInt64
	let totalSize: UInt64
}

struct TicketInspectionModel: Equatable, Sendable {
	let kind: String
	let metadata: TransferMetadataModel
}

enum ReceiverDeliveryStatus: Equatable, Sendable {
	case requested
	case accepted
	case refused
	case expired
	case completed
	case unknown
}

struct ReceiverRequestModel: Equatable, Identifiable, Sendable {
	let id: String
	let transferId: UInt64
	let remoteEndpointId: String
	let transferName: String
	let receiverName: String?
	let receiverDeviceName: String?
	let appVersion: String
	let status: ReceiverDeliveryStatus
	let reason: String?
	let requestedAt: Int64
	let respondedAt: Int64?
	let completedAt: Int64?
}

struct CoreState: Equatable, Sendable {
	var isInitialized: Bool = false
	var status: CoreStatus?
	var events: [CoreEventModel] = []
	var transfers: [Transfer] = []
	var lastShare: Share?
	var lastInspection: TicketInspectionModel?
}

/// Coalesced change hints emitted from the event sink, ported from `CoreSignal`.
enum CoreSignal: Equatable, Sendable {
	case approvalChanged(transferId: UInt64)
	case receiverHistoryChanged(transferId: UInt64)
	/// Transfer status/history changed enough to re-read the durable snapshot.
	case transfersChanged(transferId: UInt64)
}

// MARK: - Transfer helpers (ported from AppUiModels.kt)

extension TransferStatus {
	var isActiveTransfer: Bool {
		self == .importing || self == .sharing || self == .receiving
	}

	var canCancelTransfer: Bool {
		self == .importing || self == .sharing || self == .receiving
	}

	/// Terminal receive-history states eligible for deletion.
	var isTerminalReceiveHistory: Bool {
		self == .done || self == .failed || self == .cancelled
	}
}

File: shared/src/commonMain/kotlin/com/vnidrop/app/core/CoreModels.kt
package com.vnidrop.app.core

import kotlinx.coroutines.flow.SharedFlow
import kotlinx.coroutines.flow.StateFlow
import uniffi.vnidrop.ReceiveOutputSink
import uniffi.vnidrop.ReceiveOutputSinkV2

data class CoreStatus(
	val endpointId: String,
	val activeTransfers: ULong,
	val activeShares: ULong,
)

data class CoreEventModel(
	val id: String,
	val timestamp: Long,
	val scope: String,
	val transferId: ULong?,
	val direction: String?,
	val phase: String,
	val kind: String,
	val dataJson: String,
)

enum class ShareAccessPolicy {
	RequireApproval,
	AnyoneWithTransfer,
}

enum class RelayMode {
	Automatic,
	Custom,
}

data class RelaySettings(
	val mode: RelayMode = RelayMode.Automatic,
	val relayUrls: List<String> = emptyList(),
)

enum class TransferDirection {
	Send,
	Receive,
}

enum class TransferStatus {
	Importing,
	Sharing,
	Receiving,
	Done,
	Failed,
	Cancelled,
	Stopped,
}

data class Transfer(
	val localId: String,
	val transferId: ULong,
	val direction: TransferDirection,
	val status: TransferStatus,
	val peerId: String?,
	val transferName: String?,
	val contentHash: String?,
	val fileCount: ULong,
	val totalSize: ULong,
	val ticket: String?,
	val accessPolicy: ShareAccessPolicy,
	val createdAt: Long,
	val updatedAt: Long,
)

data class Share(
	val transferId: ULong,
	val ticket: String,
	val transferName: String,
	val contentHash: String,
	val fileCount: ULong,
	val totalSize: ULong,
)

data class TransferMetadataModel(
	val transferId: ULong,
	val transferName: String,
	val senderName: String?,
	val contentHash: String,
	val fileCount: ULong,
	val totalSize: ULong,
)

data class TicketInspectionModel(
	val kind: String,
	val metadata: TransferMetadataModel,
)

data class ReceiverRequestModel(
	val id: String,
	val transferId: ULong,
	val remoteEndpointId: String,
	val transferName: String,
	val receiverName: String?,
	val receiverDeviceName: String?,
	val appVersion: String,
	val status: ReceiverDeliveryStatus,
	val reason: String?,
	val requestedAt: Long,
	val respondedAt: Long?,
	val completedAt: Long?,
)

enum class ReceiverDeliveryStatus {
	Requested,
	Accepted,
	Refused,
	Expired,
	Completed,
	Unknown,
}

data class CoreState(
	val isInitialized: Boolean = false,
	val status: CoreStatus? = null,
	val events: List<CoreEventModel> = emptyList(),
	val transfers: List<Transfer> = emptyList(),
	val lastShare: Share? = null,
	val lastInspection: TicketInspectionModel? = null,
)

data class CoreStorageUsageModel(
	val blobStoreBytes: ULong,
	val databaseBytes: ULong,
	val logsBytes: ULong,
	val previewsBytes: ULong,
	val otherCoreBytes: ULong,
) {
	val appDataBytes: ULong get() = databaseBytes + logsBytes + previewsBytes + otherCoreBytes
}

data class ReceivedArtifactModel(
	val id: String,
	val locator: String,
	val locatorKind: uniffi.vnidrop.ReceivedLocatorKind,
	val logicalSize: ULong,
)

sealed interface CoreSignal {
	data class ApprovalChanged(val transferId: ULong) : CoreSignal
	data class ReceiverHistoryChanged(val transferId: ULong) : CoreSignal
	/** Transfer status/history changed enough to re-read the durable snapshot. */
	data class TransfersChanged(val transferId: ULong) : CoreSignal
}

interface CoreGateway {
	val state: StateFlow<CoreState>
	val signals: SharedFlow<CoreSignal>

	suspend fun initialize(
		appDataDir: String,
		relaySettings: RelaySettings = RelaySettings(),
	): Result<Unit>
	fun shutdown()
	suspend fun sharePath(path: String, transferName: String, senderName: String, accessPolicy: ShareAccessPolicy): Result<Share>
	suspend fun shareFileDescriptor(
		fd: Int,
		displayName: String,
		transferName: String,
		senderName: String,
		accessPolicy: ShareAccessPolicy,
	): Result<Share>
	/** Multi-source share used by multi-file pickers. */
	suspend fun shareSources(
		sources: List<uniffi.vnidrop.ShareSource>,
		transferName: String,
		senderName: String,
		accessPolicy: ShareAccessPolicy,
	): Result<Share>
	suspend fun inspectTicket(ticket: String): Result<TicketInspectionModel>
	suspend fun receive(ticket: String, outputDir: String, receiverName: String): Result<Unit>
	suspend fun receiveWithOutputSink(ticket: String, outputSink: ReceiveOutputSink, receiverName: String): Result<Unit>
	suspend fun receiveWithOutputSinkV2(ticket: String, outputSink: ReceiveOutputSinkV2, receiverName: String): Result<Unit>
	suspend fun storageUsage(): Result<CoreStorageUsageModel>
	suspend fun receivedArtifacts(): Result<List<ReceivedArtifactModel>>
	suspend fun cancel(transferId: ULong): Result<Unit>
	suspend fun delete(transferId: ULong): Result<Unit>
	suspend fun clearReceiveHistory(): Result<ULong>
	suspend fun receiverRequests(transferId: ULong): Result<List<ReceiverRequestModel>>
	suspend fun respondReceiverRequest(requestId: String, accepted: Boolean, reason: String? = null): Result<Unit>
	suspend fun refresh(): Result<Unit>
}
